### PR TITLE
Embed icons directly

### PR DESCRIPTION
**Pull request by makyo**
The following is a  to fix the fact that SVGs in most browsers don't obey viewBox attributes when included via the path of `use -> g -> svg`, but do when included via the path of `use -> svg`. To fix this, we include the icon SVG in the diagram directly, not wrapped in a `<g>`, but that means passing the icon id into `processIcon`. This specifically comes into play with the fact that the default icon on jujucharms.com is 160x160px, but icons are 96x96px, and they will not scale otherwise, but more generally, is good practice to allow icons of any size.

### CODE REVIEW

File: examples/generatesvg.go [hunk 1]
```diff
@@ -18,7 +18,7 @@ import (
 // iconURL takes a reference to a charm and returns the URL for that charm's icon.
 // In this case, we're using the api.jujucharms.com API to provide the icon's URL.
 func iconURL(ref *charm.Reference) string {
-	return "https://api.jujucharms.com/v4/" + ref.Path() + "/archive/icon.svg"
+	return "https://api.jujucharms.com/v4/" + ref.Path() + "/icon.svg"
```
Comment (makyo) [inline examples/generatesvg.go]: drive-by to use the proper endpoint

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/jujusvg/11/
Test PASSed.

**Comment by makyo** [comment]
Supersedes #26

File: svg.go [hunk 2]
```diff
@@ -34,8 +35,21 @@ func processIcon(r io.Reader, w io.Writer) error {
 		if ok && tag.Name.Space == svgNamespace && tag.Name.Local == "svg" {
 			svgStartFound = true
 			depth++
-			if err := enc.EncodeToken(tok); err != nil {
-				return errgo.Notef(err, "cannot encode token %#v", tok)
+			st := tok.(xml.StartElement)
+			for i := range st.Attr {
+				if strings.ToLower(st.Attr[i].Name.Local) == "id" {
```
Comment (rogpeppe) [inline svg.go]: Why case insensitive? I don't think XML attributes are generally considered case insensitive.

Comment (rogpeppe) [inline svg.go]: You probably want to check that Name.Space == "" too.

File: svg.go [hunk 3]
```diff
@@ -34,8 +35,21 @@ func processIcon(r io.Reader, w io.Writer) error {
...
+					st.Attr = append(st.Attr[:i], st.Attr[i+1:]...)
```
Comment (rogpeppe) [inline svg.go]: st.Attr[i].Value = id
    found = true

then append only when necessary below?

File: svg.go [hunk 4]
```diff
@@ -13,7 +14,7 @@ const svgNamespace = "http://www.w3.org/2000/svg"
 // the <svg></svg> tags, which would be invalid in this context (such as
 // <?xml...?> decls, directives, etc), writing out to a writer.  In
 // addition, loosely check that the icon is a valid SVG file.
```
Comment (rogpeppe) [inline svg.go]: It would be nice if the comment mentioned the implications and
use of the id argument.

File: svg.go (hunk 3 above)
Comment (rogpeppe) [inline svg.go]: You might want to consider defining another function, say:

   // setAttr returns the given attributes with the given attribute name set to val,
   // adding an attribute if necessary.
   func setXMLAttr(attrs []xml.Attr, name xml.Name, val string) []xml.Attr

then the original function doesn't get grubbed up so much (and you can
reuse it elsewhere)

**Comment by rogpeppe** [comment]
Looks great at first glance, with a couple of suggestions above.
I haven't looked into it in enough detail to definitively LGTM though.

File: svg.go (hunk 3 above)
Comment (makyo) [inline svg.go]: Thanks, this is a much cleaner way to go about it.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/jujusvg/12/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/jujusvg/14/
Test PASSed.

File: examples/generatesvg.go (hunk 1 above)
Comment (mhilton) [inline examples/generatesvg.go]: If you're going to use the proper endpoint it should be https://api.jujucharms.com/charmstore/v4/...

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/jujusvg/15/
Test PASSed.

File: svg_test.go [hunk 5]
```diff
@@ -137,3 +140,31 @@ func (s *SVGSuite) TestProcessIcon(c *gc.C) {
...
+			Value: "foo",
+		},
+	}
+
+	result := setXMLAttr([]xml.Attr{}, xml.Name{
+		Local: "id",
+	}, "foo")
+	c.Assert(result, gc.DeepEquals, expected)
+
+	result = setXMLAttr([]xml.Attr{
```
Comment (frankban) [inline svg_test.go]: Perhaps add another attr here to check that the extraneous attributes are not removed in the process?

**Comment by frankban** [comment]
:+1: with a single test suggestion. Is there a way to QA this?
Thank you!

**Comment by makyo** [comment]
# QA instructions

Run `go run examples/generatesvg.go examples/mediawiki-scalable.yaml > bundle.svg`.  Examine bundle.svg to ensure that icons are embedded directly in the `<defs>` section without `<g>` tags, that their id attributes are "icon-#", and that it displays properly in a browser.

File: svg.go [hunk 6]
```diff
@@ -75,3 +81,23 @@ func processIcon(r io.Reader, w io.Writer) error {
...
 }
+
+// setXMLAttr returns the given attributes with the given attribute name set to
+// val, adding an attribute if necessary.
+func setXMLAttr(attrs []xml.Attr, name xml.Name, val string) []xml.Attr {
+	found := false
+	for i := range attrs {
+		if attrs[i].Name == name {
+			attrs[i].Value = val
+			found = true
```
Comment (rogpeppe) [inline svg.go]: return attrs ?

File: svg.go [hunk 7]
```diff
@@ -75,3 +81,23 @@ func processIcon(r io.Reader, w io.Writer) error {
...
+			break
+		}
+	}
+	if !found {
```
Comment (rogpeppe) [inline svg.go]: return append(...)

?

**Comment by jrwren** [comment]
:+1:

**Comment by makyo** [comment]
Thanks, :shipit:

**Comment by makyo** [comment]
(If CI cooperates..)

**Comment by jujugui** [comment]
Status: merge request accepted. Url: http://ci.jujugui.org:8080/job/jujusvg-merge

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/jujusvg/16/
Test PASSed.

**Comment by jujugui** [comment]
Test PASSed.
Refer to this link for build results (access rights to CI server needed): 
http://ci.jujugui.org:8080//job/jujusvg/17/
Test PASSed.
